collect_app/src/main/java/org/odk/collect/android/preferences/dialogs/ResetDialogPreferenceFragmentCompat.kt

```diff
@@ -0,0 +1,270 @@
+package org.odk.collect.android.preferences.dialogs
+
+import android.content.Context
+import android.content.DialogInterface
+import android.graphics.Color
+import android.os.Bundle
+import android.view.View
+import android.widget.CompoundButton
+import androidx.appcompat.app.AlertDialog
+import androidx.appcompat.widget.AppCompatCheckBox
+import androidx.preference.PreferenceDialogFragmentCompat
+import org.odk.collect.android.R
+import org.odk.collect.android.activities.CollectAbstractActivity
+import org.odk.collect.android.fragments.dialogs.ResetSettingsResultDialog
+import org.odk.collect.android.injection.DaggerUtils
+import org.odk.collect.android.utilities.ProjectResetter
+import org.odk.collect.android.utilities.ProjectResetter.ResetAction.RESET_PREFERENCES
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.dismissDialog
+import org.odk.collect.androidshared.ui.DialogFragmentUtils.showIfNotShowing
+import org.odk.collect.async.Scheduler
+import org.odk.collect.strings.localization.getLocalizedString
+import timber.log.Timber
+import javax.inject.Inject
+
+class ResetDialogPreferenceFragmentCompat :
+    PreferenceDialogFragmentCompat(),
+    CompoundButton.OnCheckedChangeListener {
+
+    @Inject
+    lateinit var projectResetter: ProjectResetter
+
+    @Inject
+    lateinit var scheduler: Scheduler
+
+    private var preferences: AppCompatCheckBox? = null
+    private var instances: AppCompatCheckBox? = null
+    private var forms: AppCompatCheckBox? = null
+    private var layers: AppCompatCheckBox? = null
+    private var cache: AppCompatCheckBox? = null
+
+    private lateinit var _context: Context
```

Comment:
I think it doesn't make sense to keep this field you can call `getContext()` / `requireContext()` whenever you need it.

Reply:
We need this because after selecting item to reset in the dialog, we show the progress dialog so that the current dialog is detached from `context `and the `getContext()` method or `requireContext()` will return null in the `handleResult()` method.

Reply:
I think we're duplicating the pain points of `AsyncTask` here - we've ended up needing to hold onto state so that the "result" of the async work (`onPostExecute` in `AsyncTask` and the `foreground` block in `Scheduler#immediate`) can access it. If `requireContext()` returns `null` it means that the underlying Activity (the `context` for the `Fragment`) has gone along with the UI, so there is no need for us to display anything. We definitely don't want to be holding our own references to a `context` instance for async work as this can cause nasty memory leaks (Activity objects can be huge as they hold the whole view hierarchy).

The pattern to adopt here (as mentioned in [`STATE.md`](https://github.com/getodk/collect/blob/master/docs/STATE.md)) is to use `Scheduler` (coroutines under the hood) and `LiveData` to avoid needing to hold on to `Context` or views and have the view code "react" (through an `observe` call) to the result (rather than being manipulated from a task/coroutines). 

An example of this would be the use of `BlankFormsListViewModel#syncWithServer` - this fires off work in the background with `Scheduler#immediate` and then sets the resulting data on a `LiveData`. The UI code uses `observe` to react to the returned `LiveData` and we're protected from the Activity being closed etc via the lifecycle awareness features in `observe`.

For `ResetDialogPreferenceFragmentCompat`, we probably want to pull out a method (or create a `ViewModel` to host one) that returns a `LiveData` with the reset and failed actions (set from `foreground`) and the fragment can just call `handleResult` (and hide the progress dialog) from an `observe` observation.

Reply:
[ResetDialogPreferenceFragmentCompat.kt](https://github.com/getodk/collect/pull/5003/files/caee35315dd499b470d00e9cf1c52aa3246e703a#diff-be7a893a9d6744e30e4526689f3e99b0165cc2c8aecee98478345b92e22b7b0d) is dialog. When user clicks button "Reset" on this dialog, then we are showing progress dialog, running resetting in background and as result of resetting we are showing other dialog. All of this was written in the [ResetDialogPreferenceFragmentCompat.kt](https://github.com/getodk/collect/pull/5003/files/caee35315dd499b470d00e9cf1c52aa3246e703a#diff-be7a893a9d6744e30e4526689f3e99b0165cc2c8aecee98478345b92e22b7b0d)  class. Until now, we needed the `_context` variable because after pressing the "Reset" button, the dialog was dissmised, so that the `context` of dialog was null, and we used the previously saved `_context` to display the reset results. I think that was a very bad solution because like @seadowg mentioned there may be memory leak problems. Even using `LiveData` would not help because we wanted to use a `context` from view that no longer displays. 

I solved this problem by moving the responsibility of displaying the results to the fragment from which the above-mentioned dialog is shown. I also added `viewModel` to this fragment and used `LiveData` as recommended.